docker/scripts/entrypoint.sh

```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
+set -ex;
+
+# Call the initialise script
+${UNITYCATALOG_INIT_FILES}/scripts/initialise.sh
+
+
+REPO_DIR="${UNITYCATALOG_HOME}/repo"
+REPO_URL="https://github.com/unitycatalog/unitycatalog.git"
+
+
+
+[ ! -d $REPO_DIR ] && mkdir -p $REPO_DIR
+
+cd $REPO_DIR
+
+if [ -d ".git" ]; then
+    echo "Repository already cloned. Pulling latest changes..."
+    git pull
+else
+    echo "Cloning repository..."
+    git clone $REPO_URL .
```

Comment:
Why do we want to clone the repository when booting the container? Seems like an expensive operation. I would expect everything that's needed to be in there.

Reply:
This script only clones the repository once, the first time you run the container on a corresponding volume.
after cloning, the script builds the project using the `bin/start-uc-server`.

Because we're using a volume, the repository will not be cloned on each run, as long as the volume is attached.

Also the script `bin/start-uc-server` will only build the project once, because the main user "unitycatalog" has their home directory on the volume. As a result, the result of the build which produces a .cache and another sbt file, will be stored in the home directory and prevent another build from happening on the next container run.

One advantage of the above is that if you are pulling the docker image from dockerhub, you don't need to clone the GitHub repository and then copy it into the docker instance to build it.

My thinking is in the future if we create a helm chart to run in Kubernetes, we only need to point to the github repo and to the version that we need to install and everything gets set up automatically. A bit similar experience to running Airflow in Kubernetes using helm chart.

I'm certainly happy to be challenged on that, especially if someone is more experienced in this area :heart_hands:

Reply:
I believe it is more common to ship the whole application inside of the Dockerfile, instead of using `git` to pull in the application, similar to the approach in https://github.com/unitycatalog/unitycatalog/pull/22:

- The build is not deterministic: if you delete a container, and pull a new one, it might spin up a newer version of the application
- The build might take a bit of time, or might fail: Ideally you want to have everything inside of the docker container, to avoid that you're relying on pulling JARs from external services.

Reply:
@Fokko you have a really good point about the build being deterministic. In order to achieve that, we need to start adding tags on the main branch this way if we pull 1.0.0, it's always going to pull the same image.

I checked #22 and they are copying the entire repo into the docker which is quite similar to cloning the repo, unless they don't `git pull` afterwards.

Both #22 and myself are building the UC inside the docker instance, and I get your point about compiling a fat-jar on the machine and passing it to docker. If we are able to do that, than instead of using JDK on the docker machine we can use the JRE and save even more space 💡  

Let me think about it and test it. Do you think from a best practice perspective it would be a better approach ?

From a puristic perspective you want to compile any project on the target machine for the best performance but that applies to libraries in c++ more than Java. It's something for me to look up and understand a bit better when it comes to Java as java is designed to be portable.

Reply:
@jeanboutros Thanks for sharing your thoughts here. In Docker you can also compile against different architectures, most of the time this is both `amd64` and `arm64` for respectively Intel and `M[1-3]` CPU's.

> Both https://github.com/unitycatalog/unitycatalog/pull/22 and myself are building the UC inside the docker instance, and I get your point about compiling a fat-jar on the machine and passing it to docker. If we are able to do that, than instead of using JDK on the docker machine we can use the JRE and save even more space 💡

This is exactly where the multi-stage builds come in. You'll use the `JDK` to build the fat-jar. Use a `JRE` docker image to copy the jar from the `JDK` image. You discard the `JDK` image and ship the `JRE` image to dockerhub or use it locally.

This will also avoid the need of having a JDK locally.